### PR TITLE
Fix subtle corner-case with vars getting hoisted out of the loop creating unnecessary loop state

### PR DESCRIPTION
**Pull request by saipraveenb25**
+ Add tests

### CODE REVIEW

File: tests/autodiff/loop-init.slang [hunk 1]
```diff
@@ -0,0 +1,54 @@
+//TEST(compute):COMPARE_COMPUTE_EX:-slang -compute -shaderobj -output-using-type
+//TEST(compute, vulkan):COMPARE_COMPUTE_EX:-vk -compute -shaderobj -output-using-type
+
+//TEST_INPUT:ubuffer(data=[0 0 0 0], stride=4):out,name=outputBuffer
+RWStructuredBuffer<float> outputBuffer;
+
+struct A : IDifferentiable
```
Comment (csyonghe) [inline tests/autodiff/loop-init.slang:7]: Should we add some file-check test lines to ensure the intermediate context doesn't have the undesired fields?

Comment (saipraveenb25) [reply tests/autodiff/loop-init.slang:8]: Added 👍

**Review by csyonghe — APPROVED** [review]

**Review by csyonghe — APPROVED** [review]
